### PR TITLE
[TableGen] Suppress per-HwMode duplicate instructions/tables.

### PR DESCRIPTION
**Pull request by nvjle**
Currently, for per-HwMode encoding/decoding, those instructions that do not have a HwMode override are duplicated into the decoder tables for all HwModes. This includes inducing multiple tables for instructions that are otherwise unrelated (e.g., different namespace with no overrides at all).

This patch adds support to suppress instruction and table duplicates. TableGen option "-gen-disassembler --suppress-per-hwmode-duplicates" enables the suppression (off by default).

For one downstream backend with a complicated ISA and major cross-generation encoding differences, this eliminates ~32000 duplicate table entries at the time of this patch.

There are legitimate reasons to suppress or not suppress duplicates. If there are relatively few non-overridden related instructions, it can be convenient to pull them into the per-mode tables (only need to decode the per-mode tables, slightly simpler decode function in disassembler). On the other hand, in some backends, the opposite is true or the size is too large to tolerate any duplication in the first place. We let the user decide which makes sense.

This is currently off by default, though there is no reason it couldn't be enabled by default. Any existing backends downstream using the per-HwMode feature will function as before. Turning on the feature requires minor modifications to their disassembler due to more/less tables and naming.

### CODE REVIEW

**Comment by nvjle** [comment]
I hope one of @topperc @jyknight @jmolloy @0x59616e @wangpc-pp (based on recent commits or past reviews to the same general area) might review this small patch, thank you.

**Review by topperc — APPROVED** [review]
LGTM

**Comment by nvjle** [comment]
> LGTM

Thank you. Could I trouble you again to merge on my behalf?

File: llvm/test/TableGen/HwModeEncodeDecode2.td [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  let Inst{15...8} = factor;
+  let Inst{1...0} = 0b11;
+}
+
+let OutOperandList = (outs) in {
+def foo : Instruction {
+  let InOperandList = (ins i32imm:$factor);
+  let EncodingInfos = EncodingByHwMode<
+    [ModeA, ModeB], [fooTypeEncA,
+                      fooTypeEncB]
```
Comment (wangpc-pp) [inline llvm/test/TableGen/HwModeEncodeDecode2.td:51]: Indent this part.

**Review by wangpc-pp — APPROVED** [review]
LGTM

File: llvm/test/TableGen/HwModeEncodeDecode2.td (hunk 1 above)
Comment (nvjle) [reply llvm/test/TableGen/HwModeEncodeDecode2.td]: Done.

**Comment by wangpc-pp** [comment]
@nvjle You may [request a commit access](https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access) if you are going to contribute more. :-)

**Comment by nvjle** [comment]
> @nvjle You may [request a commit access](https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access) if you are going to contribute more. :-)

:100:  Definitely plan to soon-- a few more commits still needed..

**Pull request merged**